Ensure optimizer in backward works with 2d parallel

rohan-varma opened the pull request:
Summary: Test to ensure optimizer in backward works with 2D parallel.

Test Plan: CI

Differential Revision: D48508057

fduwjj commented on test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
             ]
-            optim = torch.optim.Adam(param_group, lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[0]["params"],
+                    optimizer_kwargs={"lr": param_group[0]["lr"]},
+                )
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[1]["params"],
+                    optimizer_kwargs={"lr": param_group[1]["lr"]},
+                )
```
is it possible to avoid duplicate code?

rohan-varma replied:
it can be deduped by 
```
for param_group_idx in len(param_group):
    _apply_optimizer_in_backward(
                    optimizer_class=torch.optim.Adam,
                    params=param_group[param_group_idx]["params"],
                    optimizer_kwargs={"lr": param_group[param_group_idx]["lr"]},
    )
```

but this doesn't add too much value IMO. lmk what you think

fduwjj replied:
Well, there are three places which are doing very similar thing. I do feel we can make it simpler here instead of copy paste six times.

awgu replied:
I agree that this usage of `_apply_optimizer_in_backward` is pretty verbose and ideally could be made less so.

fduwjj commented on test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
             ]
-            optim_2d = torch.optim.Adam(param_group, lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[0]["params"],
+                    optimizer_kwargs={"lr": param_group[0]["lr"]},
+                )
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[1]["params"],
+                    optimizer_kwargs={"lr": param_group[1]["lr"]},
+                )
```
ditto

fduwjj commented on test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
-            optim = torch.optim.Adam(model.parameters(), lr=0.01)
-            optim_2d = torch.optim.Adam(model_2d.parameters(), lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
```
Does this also work with DDP? If so, do you mind also testing with DDP. It can be in a separate PR. Thanks!

fegin replied:
`_apply_optimizer_in_backward` works for DDP too.

awgu reviewed (approved):
Stamping to unblock -- @fduwjj feel free to do another review

fduwjj commented on test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:
```diff
@@ -262,6 +300,13 @@ def test_2d_fsdp_integration_correctness_w_recompute_activation(self) -> None:
         self._test_2d_e2e_flow(use_orig_params=True)
 
+    @with_comms
+    @skip_if_lt_x_gpu(4)
```
yes, please use label `ciflow/periodic`

fduwjj replied:
It's defined in .ci/pytorch/multigpu-test.sh

fduwjj commented:
I don't have a strong opinion on this. If @rohan-varma wants to land this soon, I can clean it up as a follow-up PR.

rohan-varma committed:
Ensure optimizer in backward works with 2d parallel (#107748)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107748

Test to ensure optimizer in backward works with 2D parallel.

Test Plan: CI

Reviewed By: awgu

Differential Revision: D48508057

fbshipit-source-id: 7925e9076edb310e1be552135ce5b72de9a0dee8

fduwjj reviewed (approved):
LGTM

rohan-varma commented:
macos tests are unrelated
